### PR TITLE
Cybereason - Add new action Isolate Machine

### PR DESCRIPTION
**Pull request by r7-mgorny**
### Description

Describe the proposed changes:

  -

### Testing

Any testing information goes here.

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Cybereason Plugin:1.1.0. Step name: isolate_machine\n",
    "meta": {},
    "output": {
      "response": {
        "PYLUMCLIENT_INTEGRATION_CRDB18-12_020017029B0A": "Succeeded"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/cybereason:1.1.0 --debug run < tests.back/isolate_machine.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nInvalid input Pylum IDs shouldn't be empty",
    "log": "Connect: Connecting...\nrapid7/Cybereason Plugin:1.1.0. Step name: isolate_machine\nAn error occurred during plugin execution!\n\nInvalid input Pylum IDs shouldn't be empty\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.2.0-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 326, in handle_step\n    output = self.start_step(\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.2.0-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 476, in start_step\n    output = func(params)\n  File \"/usr/local/lib/python3.8/site-packages/cybereason_rapid7_plugin-1.1.0-py3.8.egg/icon_cybereason/actions/isolate_machine/action.py\", line 22, in run\n    raise PluginException(\ninsightconnect_plugin_runtime.exceptions.PluginException: An error occurred during plugin execution!\n\nInvalid input Pylum IDs shouldn't be empty\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/cybereason:1.1.0 --debug run < tests.back/isolate_machine_bad.json
</summary>
</details>

### Validate

<details>

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../tools/Makefiles/Helpers.mk ../tools/Makefiles/Colors.mk
--
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "IconValidator" failed!
        Cause: icon.png file not included in plugin.


----
[*] Total time elapsed: 1428.691ms

[*] Validating spec with js-yaml
[SUCCESS] Passes js-yaml spec check


[*] Validating markdown...
[SUCCESS] Passes markdown linting

[*] Validating python files for style...
[SUCCESS] Passes flake8 linting

[*] Validating python files for security vulnerabilities...
[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: None
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    running on Python 3.7.3
Run started:2021-02-28 19:12:14.018834

Test results:
        No issues identified.

Code scanned:
        Total lines of code: 1971
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
        Total issues (by confidence):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
Files skipped (0):
[SUCCESS] Passes bandit security checks

[*] Checking for typos with Misspell...
[SUCCESS] Passes Misspell check

```

<summary>
make validate
</summary>
</details>

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "IconValidator" failed!
        Cause: icon.png file not included in plugin.


----
[*] Total time elapsed: 16108.231ms

```

<summary>
icon-validate --all .
</summary>
</details>

### CODE REVIEW

File: cybereason/help.md [hunk 1]
```diff
@@ -39,6 +39,53 @@ Example input:
...
 
+#### Isolate Machine
+
+This action is used to isolate a machine associated with the root cause of a Malop, or to remediate a process not involved in a malop.
+
+##### Input
+
+|Name|Type|Default|Required|Description|Enum|Example|
+|----|----|-------|--------|-----------|----|-------|
+|actions_by_machine|object|None|False|Actions by machine|None|{"126811122.2298225282553311122": [{"targetId": "531122333.-3391199199911692223","actionType": "KILL_PROCESS"}]}|
```
Comment (jmcadams-r7) [inline cybereason/help.md:50]: Can we make this easier for the user? Can this object be split into separate inputs?

File: cybereason/icon_cybereason/util/api.py [hunk 2]
```diff
@@ -28,20 +28,75 @@ def connect(self):
...
+        return self.send_request(
+            "POST",
+            "/rest/sensors/action/fileSearch",
+            payload={"filters": server_filter, "fileFilters": file_filter},
+        )
+
+    def send_request(
+        self, method: str, path: str, params: dict = None, payload: dict = None
+    ) -> dict:
+        path = path.lstrip("/")
```
Comment (jmcadams-r7) [inline cybereason/icon_cybereason/util/api.py:63]: I wouldn't have done this like this. You're stripping the / here and re-adding it below in line 68. 

Do this instead: 
https://stackoverflow.com/questions/8223939/how-to-join-absolute-and-relative-urls

`urlparse.urljoin(url1, url2)`

**Review by jmcadams-r7 — CHANGES_REQUESTED** [review]

File: cybereason/plugin.spec.yaml [hunk 3]
```diff
@@ -426,4 +426,38 @@ actions:
         title: Response
         description: Search file response
         type: response
+        required: true
+  isolate_machine:
+    title: Isolate Machine
+    description: Isolate a machine associated with the root cause of a Malop, or to remediate a process not involved in a malop
```
Comment (mrinehart-r7) [inline cybereason/plugin.spec.yaml:432]:
```suggestion
    description: Isolate a machine associated with the root cause of a Malop, or to remediate a process not involved in a Malop
```

File: cybereason/plugin.spec.yaml [hunk 4]
```diff
@@ -426,4 +426,38 @@ actions:
...
+    input:
+      malop_id:
+        title: Malop ID
+        description: Malop ID to isolate a machine or empty to remediate process not involved in a malop
```
Comment (mrinehart-r7) [inline cybereason/plugin.spec.yaml:436]:
```suggestion
        description: Malop ID to isolate a machine or empty to remediate process not involved in a Malop
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

**Pull request merged**